Gems/Atom/Tools/MaterialCanvas/Code/Source/Document/MaterialCanvasDocument.cpp

```diff
@@ -373,372 +374,401 @@ namespace MaterialCanvas
+        {
+            return AZStd::string::format("%i", *v);
+        }
+        if (auto v = AZStd::any_cast<const unsigned int>(&slotValue))
+        {
+            return AZStd::string::format("%u", *v);
+        }
+        return AZStd::string();
+    }
 
-        // Perform string substitutions on a container of strings.
-        auto replaceStringsInVec =
-            [](const AZStd::string& findText, const AZStd::string& replaceText, AZStd::vector<AZStd::string>& container)
+    bool MaterialCanvasDocument::ShouldUseInstructionsFromInputNode(
+        GraphModel::ConstNodePtr outputNode, GraphModel::ConstNodePtr inputNode, const AZStd::vector<AZStd::string>& inputSlotNames) const
+    {
+        for (const auto& inputSlotName : inputSlotNames)
         {
-            for (auto& sourceText : container)
+            if (const auto slot = outputNode->GetSlot(inputSlotName))
             {
-                AZ::StringFunc::Replace(sourceText, findText.c_str(), replaceText.c_str());
+                if (slot->GetSlotDirection() == GraphModel::SlotDirection::Input)
+                {
+                    for (const auto& connection : slot->GetConnections())
+                    {
+                        if (connection->GetSourceNode() == inputNode ||
+                            connection->GetSourceNode()->HasInputConnectionFromNode(inputNode))
+                        {
+                            return true;
+                        }
+                    }
+                }
             }
-        };
+        }
+        return false;
+    }
 
-        // Slots with special types, like color, need to be converted into a type compatible with shader code.
-        auto convertSlotTypeName = [](const AZStd::string& slotTypeName) -> AZStd::string
+    AZStd::vector<AZStd::string> MaterialCanvasDocument::GetInstructionsFromSlot(
+        GraphModel::ConstNodePtr node, const AtomToolsFramework::DynamicNodeSlotConfig& slotConfig) const
+    {
+        AZStd::vector<AZStd::string> instructionsForSlot;
+
+        auto slot = node->GetSlot(slotConfig.m_name);
+        if (slot && (slot->GetSlotDirection() != GraphModel::SlotDirection::Output || !slot->GetConnections().empty()))
         {
-            if (AZ::StringFunc::Equal(slotTypeName, "color"))
+            AtomToolsFramework::CollectDynamicNodeSettings(slotConfig.m_settings, "instructions", instructionsForSlot);
+
+            // Input slots will replace the value with the name of the variable from the connected slot if one is set.
+            for (const auto& connection : slot->GetConnections())
             {
-                return "float4";
+                auto sourceSlot = connection->GetSourceSlot();
+                auto targetSlot = connection->GetTargetSlot();
+                if (sourceSlot && targetSlot && sourceSlot != targetSlot && sourceSlot != slot)
+                {
+                    ReplaceStringsInContainer(
+                        "SLOTVALUE",
```

Comment:
For the string placeholders in the template, it may be worth using a standard prefix or suffix (e.g. `_O3DE`) so that we can just write in documentation something to the effect of "identifiers ending in `_O3DE` are reserved for internal use"

Reply:
I will do a search and replace to prefix all of the tokens with O3DE_ in a dedicated, follow up PR.